File: Awful.apk/src/main/java/com/ferg/awfulapp/task/SinglePostRequest.java
package com.ferg.awfulapp.task;

import android.content.Context;
import android.net.Uri;

import com.ferg.awfulapp.constants.Constants;
import com.ferg.awfulapp.thread.AwfulThread;
import com.ferg.awfulapp.util.AwfulError;
import com.ferg.awfulapp.util.AwfulUtils;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

/**
 * Created by matt on 8/7/13.
 */
public class SinglePostRequest extends AwfulRequest<String> {

    public static final Object REQUEST_TAG = new Object();

    private String postId;
    public SinglePostRequest(Context context, String postId) {
        super(context, Constants.FUNCTION_THREAD);
        this.postId = postId;
    }


    @Override
    public Object getRequestTag() {
        return REQUEST_TAG;
    }


    @Override
    protected String generateUrl(Uri.Builder urlBuilder) {
        urlBuilder.appendQueryParameter(Constants.PARAM_ACTION, Constants.ACTION_SHOWPOST);
        urlBuilder.appendQueryParameter(Constants.PARAM_POST_ID, this.postId);
        return urlBuilder.build().toString();
    }

    @Override
    protected String handleResponse(Document doc) throws AwfulError {
        Element postbody = doc.getElementsByClass("postbody").first();
        return postbody.html();
    }

    @Override
    protected boolean handleError(AwfulError error, Document doc) {
        return error.isCritical();
    }
}

File: Awful.apk/src/main/java/com/ferg/awfulapp/task/SinglePostRequest.kt
package com.ferg.awfulapp.task

import android.content.Context
import com.ferg.awfulapp.constants.Constants.*
import com.ferg.awfulapp.preferences.AwfulPreferences
import com.ferg.awfulapp.thread.AwfulPost
import com.ferg.awfulapp.util.AwfulError
import org.jsoup.nodes.Document

/**
 * Fetch and parse a single post, using the given [postId].
 *
 * This returns the post content as raw HTML.
 */
class SinglePostRequest(context: Context, private val postId: String)
    : AwfulRequest<String>(context, FUNCTION_THREAD) {


    override val requestTag: Any
        get() = REQUEST_TAG


    init {
        with(parameters) {
            add(PARAM_ACTION, ACTION_SHOWPOST)
            add(PARAM_POST_ID, postId)
        }
    }

    override fun handleResponse(doc: Document): String {
        val prefs = AwfulPreferences.getInstance();
        val postBody = doc.selectFirst(".postbody")
        val fyadPostBody = postBody?.selectFirst(".complete_shit") ?: throw AwfulError("Couldn't find post content")
        (fyadPostBody ?: postBody).apply {
            AwfulPost.convertVideos(this, prefs.inlineYoutube)
            getElementsByTag("img").forEach {
                AwfulPost.processPostImage(
                    it,
                    false,
                    prefs
                )
            }
            getElementsByTag("a").forEach(AwfulPost::tryConvertToHttps)
            if (this == fyadPostBody) {
                // FYAD sigs are currently a sibling div alongside .complete_shit, so we need to stick them at the end of the content
                postBody.selectFirst("> .signature")?.appendTo(this)
            }
            return html()
        }
}

    companion object {
        private val REQUEST_TAG = Any()
    }
}
